server: retry store id allocation on the correct error

andreimatei opened the pull request:
see individual commits

andreimatei committed:
kv: rename DistSQLRetryableError to UnhandledRetryableError

This recognizes that DistSQLRetryableError were seen by layers above the
TxnCoordSender for non-transactional request. The previous thinking had
been that the TxnCoordSender converts all lower-level retryable errors
into HandledRetryableError, but that wasn't the case for
non-transactional requests.

This also adds a test for retryable errors in non-transactional
requests.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15884)
<!-- Reviewable:end -->

andreimatei commented:
test forthcoming

a-robinson reviewed (approved):
The code change LGTM, I'll check out the test once you add it.

a-robinson commented:
This might fix the few node ID allocation panics we've seen in Sentry - https://sentry.io/cockroach-labs/cockroachdb/issues/273211238/

bdarnell commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 866 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15884#-KkBwkA0_Xcr3bcQ12UJ:-KkBwkA1C3aeK0NBTDqQ:b-ke28bs) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec2f90af5cd8718d1ffa4bd243ac566561d9a25e/pkg/kv/txn_coord_sender.go#L866)):*
> ```Go
> 				// HandledRetryableError.
> 				pErr = roachpb.NewError(
> 					roachpb.NewHandledRetryableTxnError(pErr.Message, nil, roachpb.Transaction{}))
> ```

We haven't "handled" the error in any meaningful way, so maybe we should just make the higher levels tolerate unhandled retryable errors?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15884)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 866 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15884#-KkBwkA0_Xcr3bcQ12UJ:-KkBxZBNXkh2i6IgbXBL:b-s4meib) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec2f90af5cd8718d1ffa4bd243ac566561d9a25e/pkg/kv/txn_coord_sender.go#L866)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

We haven't "handled" the error in any meaningful way, so maybe we should just make the higher levels tolerate unhandled retryable errors?
</blockquote></details>

the problem, described in the commit msg, is that, without this "handling", higher levels will see a `DistSQLRetryableError` (produced by `pErr.GoError()`).
We could rename `DistSQLRetryableError` to something else. Or we could have `pErr.GoError()` return the error detail directly - but the problem with that is that, if you wrap the `ErrorDetail` in a `pErr` again you'd get a retryable error that might restart the wrong txn.

I'd rather just do what I'm doing here... No?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15884)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 866 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15884#-KkBwkA0_Xcr3bcQ12UJ:-KkBzkPM-g7SS38qFmqh:buks1qi) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec2f90af5cd8718d1ffa4bd243ac566561d9a25e/pkg/kv/txn_coord_sender.go#L866)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

the problem, described in the commit msg, is that, without this "handling", higher levels will see a `DistSQLRetryableError` (produced by `pErr.GoError()`).
We could rename `DistSQLRetryableError` to something else. Or we could have `pErr.GoError()` return the error detail directly - but the problem with that is that, if you wrap the `ErrorDetail` in a `pErr` again you'd get a retryable error that might restart the wrong txn.

I'd rather just do what I'm doing here... No?
</blockquote></details>

DistSQLRetryableError should be renamed now that the assumption that only distsql sends non-transactional requests has been shown to be false. We should clean this up instead of compounding the issues by returning "handled" errors to avoid returning DistSQLRetryableError to non-distsql callers. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15884)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 20 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/kv/txn_coord_sender.go, line 866 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15884#-KkBwkA0_Xcr3bcQ12UJ:-KkCeVXI1alSL_XWbajP:b8e3hl3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec2f90af5cd8718d1ffa4bd243ac566561d9a25e/pkg/kv/txn_coord_sender.go#L866)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

DistSQLRetryableError should be renamed now that the assumption that only distsql sends non-transactional requests has been shown to be false. We should clean this up instead of compounding the issues by returning "handled" errors to avoid returning DistSQLRetryableError to non-distsql callers. 
</blockquote></details>

I've added a commit removing the "handled" and the "txn" part of the name - there's now a `roachpb.RetryableError`.
I've left the `DistSQLRetryableError` alone.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15884)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 20 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/kv/txn_coord_sender.go, line 866 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15884#-KkBwkA0_Xcr3bcQ12UJ:-KkD0suNv9T8fmXSg4si:b-2f30au) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec2f90af5cd8718d1ffa4bd243ac566561d9a25e/pkg/kv/txn_coord_sender.go#L866)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I've added a commit removing the "handled" and the "txn" part of the name - there's now a `roachpb.RetryableError`.
I've left the `DistSQLRetryableError` alone.
</blockquote></details>

Removing the "handled" part makes sense, but I don't think removing "txn" does - this is inherently a transaction-related error. The one thing it does is add a new transaction proto to an existing error. Therefore I don't think it makes sense to use it here to attach an empty Transaction proto to a normal error just to satisfy an overly-narrow assertion.

I stand by my suggestion to make the higher-level code (the retry loop in `node.go`) more tolerant of unexpected error types. It should default to retrying instead of failing - it should probably retry unconditionally, as `storage/id_alloc.go` does. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15884)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 20 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/kv/txn_coord_sender.go, line 866 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15884#-KkBwkA0_Xcr3bcQ12UJ:-KkDG8RbznQMHfvvC7VN:b2tkyib) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec2f90af5cd8718d1ffa4bd243ac566561d9a25e/pkg/kv/txn_coord_sender.go#L866)):*
>  The one thing it does is add a new transaction proto to an existing error. Therefore I don't think it makes sense to use it here to attach an empty Transaction proto to a normal error just to satisfy an overly-narrow assertion.

That's true enough, although that proto had been commented to not be used by non-transactional requests.

Regardless of what errors the idallocator retries on, it shouldn't see an error called `DistSQLRetryableTxnError`. Should we name that one... `StorageRetryableError` ?
To recap, that one is produced by `pErr.GoeError()` and it needs to be a proto that DistSQL sends over the wire. The alternative is to have `GoError()` return the `ErrorDetail` directly, but as I was saying above I personally don't like that very much.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15884)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 20 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/kv/txn_coord_sender.go, line 866 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15884#-KkBwkA0_Xcr3bcQ12UJ:-KkHQDQ31actZjFtgqdz:b9fmmvu) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec2f90af5cd8718d1ffa4bd243ac566561d9a25e/pkg/kv/txn_coord_sender.go#L866)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

>  The one thing it does is add a new transaction proto to an existing error. Therefore I don't think it makes sense to use it here to attach an empty Transaction proto to a normal error just to satisfy an overly-narrow assertion.

That's true enough, although that proto had been commented to not be used by non-transactional requests.

Regardless of what errors the idallocator retries on, it shouldn't see an error called `DistSQLRetryableTxnError`. Should we name that one... `StorageRetryableError` ?
To recap, that one is produced by `pErr.GoeError()` and it needs to be a proto that DistSQL sends over the wire. The alternative is to have `GoError()` return the `ErrorDetail` directly, but as I was saying above I personally don't like that very much.
</blockquote></details>

It's not a `StorageRetryableError` - package `storage` never sees these. Maybe `RetryableGoError` (and put `Handled` back on the `ErrorDetail`)? We've painted ourselves into a corner with this design; all the options are kind of ugly. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15884)*
<!-- Sent from Reviewable.io -->

andreimatei committed:
server: retry store id allocation on the correct error

Fixes #15870

The code was retrying on an error only meant for transactional requests,
even though it was performing non-transactional requests.

andreimatei commented:
Review status: 0 of 11 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/kv/txn_coord_sender.go, line 866 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15884#-KkBwkA0_Xcr3bcQ12UJ:-KkHoJAoaMJ3bt3dazQ9:b-im403d) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec2f90af5cd8718d1ffa4bd243ac566561d9a25e/pkg/kv/txn_coord_sender.go#L866)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

It's not a `StorageRetryableError` - package `storage` never sees these. Maybe `RetryableGoError` (and put `Handled` back on the `ErrorDetail`)? We've painted ourselves into a corner with this design; all the options are kind of ugly. 
</blockquote></details>

Switched to `HandledRetryableTxnError` and `UnhandledRetryableError`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15884)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 10 of 11 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15884#-:-KkI-weT7VbFOZlvFliQ:bnfp4nl)*
<!-- Sent from Reviewable.io -->
